structuretoolkit/analyse/snap.py

```
@@ -0,0 +1,601 @@
+    Internal helper function to set LAMMPS variables
+
+    Args:
+        lmp (lammps.Lammps): Lammps library instance
+        **lmp_variable_args (dict): key value pairs of LAMMPS variables to set
+    """
+    for k, v in lmp_variable_args.items():
+        lmp.command(f"variable {k} equal {v}")
+
+
+def _set_computes_snap(lmp, bispec_options):
+    """
+    Set LAMMPS computes to calculate SNAP descriptors
+
+    Args:
+        lmp (lammps.Lammps): Lammps library instance
+        bispec_options (dict): bi-spectrum component settings
+    """
+    # # Bispectrum coefficient computes
+    base_b = "compute b all sna/atom ${rcutfac} ${rfac0} ${twojmax}"
+    base_db = "compute db all snad/atom ${rcutfac} ${rfac0} ${twojmax}"
+    base_vb = "compute vb all snav/atom ${rcutfac} ${rfac0} ${twojmax}"
+
+    numtypes = bispec_options["numtypes"]
+    radelem = " ".join([f"${{radelem{i}}}" for i in range(1, numtypes + 1)])
+    wj = " ".join([f"${{wj{i}}}" for i in range(1, numtypes + 1)])
+
+    kw_options = {
+        k: bispec_options[v]
+        for k, v in {
+            # "diagonal": "diagonalstyle", For legacy diagonalstyle option
+            "rmin0": "rmin0",
+            "bzeroflag": "bzeroflag",
+            "quadraticflag": "quadraticflag",
+            "switchflag": "switchflag",
+        }.items()
+        if v in bispec_options
+    }
+    kw_substrings = [f"{k} {v}" for k, v in kw_options.items()]
+    kwargs = " ".join(kw_substrings)
+
+    for op, base in zip(("b", "db", "vb"), (base_b, base_db, base_vb)):
+        command = f"{base} {radelem} {wj} {kwargs}"
+        lmp.command(command)
+
+    for cname in ["b", "db", "vb"]:
+        lmp.command(f"compute {cname}_sum all reduce sum c_{cname}[*]")
+
+
+def _extract_computes_snap(lmp, num_atoms, n_coeff, num_types):
+    lmp_atom_ids = lmp.numpy.extract_atom_iarray("id", num_atoms).flatten()
+    assert np.all(
+        lmp_atom_ids == 1 + np.arange(num_atoms)
+    ), "LAMMPS seems to have lost atoms"
```

There exists a condition under which the [assert will get skipped](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) -- are you sure you want this and not a try/except?

---

The assert statement is a way to fail early. If users run their python code with `-O` to improve performance they do not get this nice error message but rather a more cryptic one when we try to read from the LAMMPS memory. So from my perspective it is a hint when things go wrong rather than a try/except condition.